Issue 1075: Add ETL tests to all data classes

lucasmbrown-usds opened the pull request:
#1075 . Work in progress by @saran-ahluwalia and myself to add tests to all ETL classes. 

Out of the punchlist [here](https://github.com/usds/justice40-tool/issues/1075#issuecomment-1001700157), so far we only have finished updating NRI. 

@esfoobar-usds @widal001  - let's talk about the overall approach. If this looks good to you, we can add these tests to every other ETL class. Let's get an approach we're happy with in one place before duplicating it elsewhere, though.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/etl/base.py:
```diff
@@ -107,7 +107,10 @@ def extract(
         if source_url and extract_path:
             unzip_file_from_url(
-                source_url, self.TMP_PATH, extract_path, verify=verify
+                file_url=source_url,
```
Just adding some named arguments because I love them.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/data/acs.csv:
```diff
@@ -1,11 +0,0 @@
-GEOID10,POPULATION
```
This file is no longer needed because we no longer convert from tracts to CBGs.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/data/input.csv:
```diff
@@ -1,11 +1,11 @@
 OID_,NRI_ID,STATE,STATEABBRV,STATEFIPS,COUNTY,COUNTYTYPE,COUNTYFIPS,STCOFIPS,TRACT,TRACTFIPS,POPULATION,BUILDVALUE,AGRIVALUE,AREA,RISK_SCORE,RISK_RATNG,RISK_NPCTL,RISK_SPCTL,EAL_SCORE,EAL_RATNG,EAL_NPCTL,EAL_SPCTL,EAL_VALT,EAL_VALB,EAL_VALP,EAL_VALPE,EAL_VALA,SOVI_SCORE,SOVI_RATNG,SOVI_NPCTL,SOVI_SPCTL,SOVI_VALUE,RESL_SCORE,RESL_RATNG,RESL_NPCTL,RESL_SPCTL,RESL_VALUE,AVLN_EVNTS,AVLN_AFREQ,AVLN_EX ...
-1,T15007040300,Hawaii,HI,15,Kauai,County,007,15007,040300,15007040300,8385,912658000.000000000000000,147860.564720087801106,3.610852158936000,18.070583080265969,Relatively Low,63.077578740428109,63.496932515337420,18.619940187472693,Relatively Low,59.642007726254143,70.552147239263803,324935.215571426844690,98076.524868236796465,0.029679044196817,225560.735895809746580,1297.954807380313696,31.680 ...
+1,T15007040300,Hawaii,HI,15,Kauai,County,007,15007,040300,15007040300,8385,992658000.000000000000000,147860.564720087801106,3.610852158936000,18.070583080265969,Relatively Low,63.077578740428109,63.496932515337420,18.619940187472693,Relatively Low,59.642007726254143,70.552147239263803,324935.215571426844690,98076.524868236796465,0.029679044196817,225560.735895809746580,1297.954807380313696,31.680 ...
```
Changed two rows of input data as an example.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/data/transform.csv:
```diff
@@ -1,56 +1,56 @@
 OID_,NRI_ID,STATE,STATEABBRV,STATEFIPS,COUNTY,COUNTYTYPE,COUNTYFIPS,STCOFIPS,TRACT,GEOID10_TRACT,POPULATION,BUILDVALUE,AGRIVALUE,AREA,RISK_SCORE,RISK_RATNG,RISK_NPCTL,RISK_SPCTL,FEMA Risk Index Expected Annual Loss Score,EAL_RATNG,EAL_NPCTL,EAL_SPCTL,EAL_VALT,EAL_VALB,EAL_VALP,EAL_VALPE,EAL_VALA,SOVI_SCORE,SOVI_RATNG,SOVI_NPCTL,SOVI_SPCTL,SOVI_VALUE,RESL_SCORE,RESL_RATNG,RESL_NPCTL,RESL_SPCTL,RES ...
-1,T15007040300,Hawaii,HI,15,Kauai,County,7,15007,40300,15007040300,8385,912658000.0000000000,147860.5647200878,3.6108521589,18.0705830803,Relatively Low,63.0775787404,63.4969325153,18.6199401875,Relatively Low,59.6420077263,70.5521472393,324935.2155714268,98076.5248682368,0.0296790442,225560.7358958097,1297.9548073803,31.6808724993,Relatively Moderate,48.7278745931,51.8518518519,-0.1330000000,52. ...
```
This is all changed because we stopped forcing a certain number of decimals in the output. All the numbers have stayed the same where we expected them to, though.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/test_etl.py:
```diff
@@ -10,8 +14,41 @@
 )
 
+"""
+Set UPDATE_TEST_FIXTURES to `True` temporarily to allow you to quickly update the test
```
We'll put this in a README somewhere too.

widal001 commented:
@lucasmbrown-usds @saran-ahluwalia Thanks for pushing this work forward! I'll aim to take a look at this tonight and comment on this thread

widal001 reviewed (commented):
### Summary

@lucasmbrown-usds @saran-ahluwalia These look like solid improvements and the existing unit tests run successfully with the changes, so they should be good to merge. I added specific suggestions in inline comments and offered some broader considerations for next steps for this strategy below, but they shouldn't be blockers for this PR.

### Highlights

- Love the move to standardize unit tests across the ETL classes and these changes are a solid foundation to build on
- Automating the method to update the test fixtures for each ETL source is a smart way to save time and effort, if approached with the appropriate caution when populating those files
- Appreciate the introduction of a test for the `extract()` method that was an oversight in the initial testing suite that I wrote

### Considerations

- While I like the approach of automating the generation/updating of test fixtures, we'll just need to be careful we don't codify any bugs in the fixtures, since it seems like the ETL code being tested against those files is the same code used to generate them (at least in the transform step) 
- This is a bit outside the scope of the changes introduced by this PR, but something I flagged when submitting the first test suite for the NRI class is that the two different `conftest.py` (the one in `data_pipeline/tests/` and the one in `data_pipeline/ETL/tests/`) actually write tmp files to different directories. As a result, depending on which subset of the tests we're running, the location of the temp files created by tests can change. This doesn't result in any failures but can make it confusing if you're trying to manually inspect the output after executing the tests. I can explain why this is happening if it would be helpful, though it might be easier to explain in real-time instead of asynchronously. 
- It may be worth validating that the output test fixture adheres to the contract established for the output of each ETL source class. That was the initial goal behind the work done in #619 that I regrettably didn't get a chance to expand on. The `ExtractTransformLoad.validate_output()` method will likely need to be adjusted if we no longer use CBG as the common geographic unit for source data, but an updated version could be useful to ensure that the fixtures being used for each testing suite adhere to a standard contract that the score classes expect.

### Suggestions

- If the goal is to replicate a variation of this testing suite across each of the ETL source classes, it may be beneficial to create a `TestBaseETL` class with common code for each UUT in the ETL source class. That base test class can then be inherited by the respective test classes for each ETL source class per [this issue](https://stackoverflow.com/questions/52044490/replacement-for-test-case-inheritance-in-pytest)
- Abstracting a base class for ETL source tests might also be easier if we were to update the ETL source classes to utilize some of the standardized class attributes also introduced in #619, for example using the following items set in `ExtractTransformLoad.__init__()`. Admittedly, that probably isn't a requirement for getting unit testing in place and may be outside the scope of this immediate goal, even if it could help simplify how these tests are written.
```python
def __init__(self, config_path: Path) -> None:
        """Inits the class with instance specific variables"""

        # set by _get_yaml_config()
        self.NAME: str = None
        self.SOURCE_URL: str = None
        self.GEOID_COL: str = None
        self.GEO_LEVEL: str = None
        self.SCORE_COLS: list = None
        self.FIPS_CODES: pd.DataFrame = None
        self.OUTPUT_PATH: Path = None
        self.CENSUS_CSV: Path = None

        self._get_yaml_config(config_path)
```

lucasmbrown-usds commented:
In reponse to your excellent and thoughtful review, @widal001 -- 

> This is a bit outside the scope of the changes introduced by this PR, but something I flagged when submitting the first test suite for the NRI class is that the two different conftest.py (the one in data_pipeline/tests/ and the one in data_pipeline/ETL/tests/) actually write tmp files to different directories

Great catch. I created https://github.com/usds/justice40-tool/issues/1092 to track that. 

> It may be worth validating that the output test fixture adheres to the contract established for the output of each ETL source class.

I actually had never noticed `validate_output` on the base class, that's super cool. Question: do you think it's smarter to run that as part of the test, or as part of each ETL (e.g., make the ETL runner call `extract`, `transform`, `load`, and then `validate_output`)? Or maybe in both places! 

> TestBaseETL

Yeaaaa I really like that idea. That would clean this up a lot and it would also be nice to have explicit test contracts on the parent class that make it obvious that all child classes must pass these same tests, in addition to the child-specific tests. 

I think sadly the `_get_yaml_config` is a relic from a different vision at a different time -- we've moved to specifying all those parameters directly in the code now. But I do like the idea of forcing standard parameters across classes to simplify things and forcing checks like, does it have a tract ID if the `GEO_LEVEL` is `tract`, or a CBG ID if the `GEO_LEVEL` is `CBG`. 

Some of the things in those suggested parameters I've been considering as part of https://github.com/usds/justice40-tool/issues/1013 too. This is good motivation to start down that path! 

(more responses to come)

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/test_etl.py:
```diff
@@ -36,6 +73,54 @@ def test_init(self, mock_etl, mock_paths):
         assert etl.GEOID_TRACT_FIELD_NAME == "GEOID10_TRACT"
 
+    @mock.patch("data_pipeline.utils.requests")
+    def test_extract(self, requests_mock, mock_paths):
```
Yeaaa good catch! To be honest I am much more familiar with this individualized way of doing a `@mock.patch` on the method, but it is out-of-sync with how the rest of the test fixtures work and I shouldl take the extra step to learn something new today and bring more consistency to the codebase. 

Thanks for flagging this, I will refactor!

widal001 replied:
@lucasmbrown-usds After reflecting on this further, I realized that if we're abstracting these tests to a base test class (`TestETLBase`) then there probably won't be much added benefit of promoting this segment of code to a pytest fixture since other ETL classes would simply use `super().test_extract()` instead of the `requests_mock` directly. I'm all for learning new stuff, haha, but it seems like you could de-scope this from the immediate refactor tasks you outlined below

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/test_etl.py:
```diff
@@ -10,8 +14,41 @@
 
 class TestNationalRiskIndexETL:
+    def test_update_test_fixtures(self, mock_paths, mock_etl):
+        """Assert that UPDATE_TEST_FIXTURES is False."""
```
Overall these are really great flags, thanks. And I appreciate your care for detail and the extremely appropriate paranoia about what a future distracted developer might do with this! 

A few notes. 

Overall my philosophy here was to try to replicate the philosophy of [snapshot testing](https://blog.bitsrc.io/whats-snapshot-testing-and-can-we-use-it-for-frontend-f5a441cae27a) from the front-end world for backend ETL scripts. (Now that I'm talking about it, I realize that I 100% stole this idea from our friend @dfkoh !) Front-end snapshots will update with a single button press, and thus indeed it's true that the code under test will create the updated test output. 

With front-end snapshot testing, the approach tries to balance two needs: 

1. Alert the developer that their code has caused something to change in the site rendering, which may be intentional or unintentional. 
2. Make the tests easy enough to update that developers actually maintain good code coverage. 

In addition, by making the changes in the snapshot part of the codebase, any changes in outputs are subject to peer review. 

While updating the NRI tests for various small changes, I found myself getting fairly tired of how much work it takes to update the fixtures. (We have a similar challenge with https://github.com/usds/justice40-tool/issues/1048.) 

I'd love to have a way that easily updates the fixtures *after* I've carefully reviewed them to make sure that the outputs are indeed as expected. This also shows up in the `git diff` for a pull request. But as with front-end testing, to quote that link above, "At first many developers tend to blindly update the snapshots when they come across a failing test. However, blindly updating snapshots should never be done." 

I like the idea of making this a common method! And also maybe it shouldn't be a `test_` method at all, but instead a non-test method that gets run explicitly by the developer. Hmm.

vim-usds replied:
@lucasmbrown-usds - strongly agree on this:

> "At first many developers tend to blindly update the snapshots when they come across a failing test. However, blindly updating snapshots should never be done."

Given how simple it is to update snapshots on the FE, under deadlines, I am guilty of this (on other projects). However reading failures make take a few extra minutes and worth the extra time to understand why it's failing rather than blindly updating all snapshots.

I also would say that updating ALL snapshots versus updating the one that is failing is also a good practice.

widal001 replied:
This is such a cool strategy! Doesn't surprise me that snapshot tests for backend code is David's brainchild haha. I'll definitely have to adopt them as a part of my testing toolkit.

lucasmbrown-usds commented:
> * Update validate_output and make sure it runs either as part of the tests or as part of the ETL itself

FYI I'm going to punt this to https://github.com/usds/justice40-tool/issues/1186.

widal001 commented:
@lucasmbrown-usds Sorry I missed this question earlier!

> I actually had never noticed `validate_output` on the base class, that's super cool. Question: do you think it's smarter to run that as part of the test, or as part of each ETL (e.g., make the ETL runner call `extract`, `transform`, `load`, and then `validate_output`)? Or maybe in both places!

Eventually this would be ideal to have as a part of the ETL following `transform` so it could be called by the ETL runner but I was anticipating excluding it for now to avoid breaking the existing ETL workflows that might not adhere to the contract conditions tested in `validate_output()`

As a bridge to updating the existing ETL classes, it would be great to include it in the tests. In particular I think it would be crucial to include in the refactored `update_test_fixtures()` because it could ensure that the fixtures we're testing the ETL against adhere to that contract. We can do a deeper dive at 2pm to walk through some ideas for incorporating `validate_output()` in the work you've done on that function

lucasmbrown-usds commented:
FYI @emma-nechamkin @widal001 -- I've implemented some of the code for updating test fixtures. 

I created an ExampleETL class used by the base ETL test class. All it does is unzip a CSV and multiply the column by 2. I updated a cell in the "input.zip" fixture (from 3.1 to 6.1), and then the following rows updated in the 3 "downstream" CSVs in this nice clean diff. 🎉 

<img width="1156" alt="Screen Shot 2022-01-27 at 4 50 33 PM" src="https://user-images.githubusercontent.com/25443529/151450060-7e3b662d-cd76-4083-bdc8-6c36d0c49c09.png">

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/base/config.yaml:
```diff
@@ -1,10 +0,0 @@
-name: Template
```
These files are all from an older approach to dataset schemas that's out of date.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/conftest.py:
```diff
@@ -22,8 +22,9 @@ def copy_data_files(src: Path, dst: Path) -> None:
-        copyfile(src, dst)
-        assert dst.exists()
+
+    copyfile(src, dst)
```
This was a small but important bug! The data was only copied if the destination file didn't exist.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/conftest.py:
```diff
@@ -34,24 +35,17 @@ def mock_paths(tmp_path_factory) -> tuple:
-
 
-@pytest.fixture(scope="session")
-def mock_census(mock_paths) -> Path:
```
No longer needed.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/example/etl.py:
```diff
@@ -0,0 +1,58 @@
+
+        df[self.EXAMPLE_FIELD_NAME] = df[self.INPUT_FIELD_NAME] * 2
+
+        self.output_df = df
```
`load` is now implemented on the base class, so nothing more needed here. 🎉 !

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/test_etl.py:
```diff
@@ -1,17 +1,58 @@
+    def _setup_etl_instance_and_run_extract(
+        self, mock_etl, mock_paths
+    ):
+        with mock.patch('data_pipeline.utils.requests') as requests_mock:
```
moved this from being a decorator to being a context manager so that the signature of `_setup_etl_instance_and_run_extract` is identical between the base and child classes. 

Also sadly @widal001 I don't think this makes sense to move to conftest.py, because we're monkeypatching `requests` in a very specific way that's going to return the NRI data, so I don't think it's useful on other classes. Do you agree?

emma-nechamkin commented on data/data-pipeline/data_pipeline/tests/sources/test_etl_base.py:
```diff
@@ -0,0 +1,244 @@
+        assert etl.output_df.shape[0] > 0
+
+        # Check that it has all columns
+        for col in etl.COLUMNS_TO_KEEP:
```
I think if there's no uniqueness test for the columns, it might make sense to check shape[1] with the idea that having every column and having every column precisely one time are not the same.

lucasmbrown-usds replied:
I added a uniqueness test on `COLUMNS_TO_KEEP` - let me know if you still think we should update this.

emma-nechamkin replied:
I think the uniqueness test is sufficient!

emma-nechamkin commented on data/data-pipeline/data_pipeline/tests/sources/test_etl_base.py:
```diff
@@ -0,0 +1,244 @@
+
+        # check that the `COLUMNS_TO_KEEP` are in the output
+        for col in etl.COLUMNS_TO_KEEP:
+            assert col in actual_output.columns, f"{col} is missing from output"
```
Same point about uniqueness, feel free to disregard

emma-nechamkin reviewed (commented):
This is quite clear and I think comprehensive! Small nits only.

vim-usds commented:
Assumption here is that this is for the first ETL and then we will expand to others

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/test_etl.py:
```diff
@@ -26,15 +66,86 @@ def test_init(self, mock_etl, mock_paths):
+            dst=self._DATA_DIRECTORY_FOR_TEST / self._OUTPUT_CSV_FILE_NAME,
+        )
+
+    def test_extract(self, mock_etl, mock_paths):
```
good call - apparently i hadn't!

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/test_etl.py:
```diff
@@ -76,23 +189,34 @@ def test_load(self, mock_etl):
-        assert output_path.exists()
-        assert output.shape == (55, 5)
-        pd.testing.assert_frame_equal(output, expected, rtol=1e-10, atol=1e-10)
+        assert actual_output.shape == (55, 5)
```
Yeah, good point. This is redundant. I'll remove. That also means we can probably move this to the base class!

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/test_etl_base.py:
```diff
@@ -0,0 +1,244 @@
+
+        However, note a few things first:
+
+        1. Do *not* update these fixtures if you did not expect the ETL results to
```
Right right. Let's have that conversation over slack or in person at some point -- I could be persuaded either the "one flag per fixture" or "one flag all fixtures" route.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/test_etl_base.py:
```diff
@@ -0,0 +1,244 @@
+        assert etl.SOURCE_URL is not None
+        assert etl.GEO_LEVEL is not None
+        assert etl.COLUMNS_TO_KEEP is not None
+        assert len(etl.COLUMNS_TO_KEEP) > 0
```
love it!

lucasmbrown-usds commented:
> Assumption here is that this is for the first ETL and then we will expand to others

Indeed. I created #1217 with a checklist (that we should try out and find out what it's missing!) for each other ETL class.

emma-nechamkin commented on data/data-pipeline/data_pipeline/etl/base.py:
```diff
@@ -27,72 +36,68 @@ class ExtractTransformLoad:
     # TODO: investigate. Census says there are only 74,134 tracts in the US,
     #  Puerto Rico, and island areas. This might be from tracts at different time
     #  periods. https://github.com/usds/justice40-tool/issues/964
     EXPECTED_MAX_CENSUS_TRACTS: int = 74160
```
very annoying QQ -- wondering if we can set this in a constants file rather than in the class itself, with the idea that if the data changes, it will be less annoying to change? no strong feelings though.

lucasmbrown-usds replied:
Good idea. Created this to track: https://github.com/usds/justice40-tool/issues/1245

emma-nechamkin commented on data/data-pipeline/data_pipeline/etl/base.py:
```diff
@@ -116,63 +124,150 @@ def transform(self) -> None:
+                geo_field_character_lengths = (
+                    self.output_df[geo_field].str.len().unique()
+                )
+
```
I think using `nunique` would be clearer, but would contain less information?

Like, `assert self.output_df[geo_field].str.len().nunique() == 1` and then `len(self.output_df[geo_field].values[0]) == `expected` is more legible to me personally, but again, no strong preference here.

lucasmbrown-usds replied:
Makes sense! Updated.

emma-nechamkin commented on data/data-pipeline/data_pipeline/etl/base.py:
```diff
@@ -116,63 +124,150 @@ def transform(self) -> None:
+                        "has the wrong length."
+                    )
+
+                non_unique_geo_field_values = len(
```
I think this would be clearer with `nunique` but it's a personal preference

```
self.output_df[geo_field].shape[0] == self.output_df[geo_field].nunique()
```

emma-nechamkin commented on data/data-pipeline/data_pipeline/etl/runner.py:
```diff
@@ -60,6 +60,9 @@ def etl_runner(dataset_to_run: str = None) -> None:
         etl_instance.load()
 
+        # run validate
+        etl_instance.validate()
```
🥇 yay

emma-nechamkin commented on data/data-pipeline/data_pipeline/tests/sources/national_risk_index/test_etl.py:
```diff
@@ -26,73 +96,56 @@ def test_init(self, mock_etl, mock_paths):
+            etl.EXPECTED_POPULATION_LOSS_RATE_FIELD_NAME,
+            etl.EXPECTED_AGRICULTURE_LOSS_RATE_FIELD_NAME,
+            etl.EXPECTED_BUILDING_LOSS_RATE_FIELD_NAME,
+        ]
```
I am a little confused here from the perspective of updating / changing data -- like it would be really nice imo if all of this were housed in a single place outside of the test, with the idea that it could be shared by the etls themselves and also the tests -- but it's entirely possible i'm not understanding something here :)

lucasmbrown-usds replied:
Oh I see -- your idea would be like, maybe there's a configuration file or something where we specify this, and then we're loading that configuration from both the ETLs and the tests? 

That is fairly reasonable to me (and not far off from the ideas of the Big Refactor). At the same time, with unit tests, it's really good to actually avoid being DRY (see this [post](https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html), or the principle of having tests be ["WET"](https://www.heroku.com/podcasts/codeish/102-whether-to-repeat-yourself-dry-damp-or-wet) (write everything twice). 

So I appreciate that if someone needs to update the columns to keep, that the test will explicitly fail and they have to manually update it. This is can catch an accidental modification of COLUMNS_TO_KEEP that would get loaded into the test from the same config and pass. 

What do you think?

emma-nechamkin replied:
That seems reasonable. It's slightly at odds with the "update all pickles!" flag but :shrug: probably in a good way. I'm convinced!

emma-nechamkin reviewed (commented):
I don't have any substantive comments -- this looks great to me!

esfoobar-usds reviewed (approved):
This looks good to me. The only small nit is around clarity of when there should be pickles generated and when not, but let's see how it goes with the next tests we write.

🍏

lucasmbrown-usds commented:
> The only small nit is around clarity of when there should be pickles generated and when not

Yeah right now we're comparing only against the values of the dataframes, not full pickles which would also allow us to compare types (but also, pickles introduce some challenges around the version of pandas used to create the files). I think the main downside of switching to pickles would be that the diffs in the snapshot updates would no longer be human-readable and would make oversight of those changes a lot harder.

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/etl/score/etl_score.py:
```diff
@@ -111,17 +114,7 @@ def extract(self) -> None:
-            dtype={self.GEOID_TRACT_FIELD_NAME: "string"},
-            low_memory=False,
-        )
+        self.national_risk_index_df = NationalRiskIndexETL.get_data_frame()
```
I agree! I've been dreaming of this day 🤓

widal001 reviewed (commented):
## Summary

@lucasmbrown-usds This all looks solid! Most of the comments I left were for forward-looking considerations beyond the scope of this PR. 

The only immediate consideration before merging is to rename `TestETL.test_update_fixtures()` to `TestETL.update_fixtures()` to avoid accidentally updating the fixtures in the ETL test classes that inherit from when `test_update_fixtures()` isn't re-implemented correctly in those classes. This change would prevent pytest from executing that code unless a derivative test specifically implemented `test_update_fixtures()` and used a `super().update_fixtures()` to invoke that method.

## Other Thoughts/Suggestions

While these thoughts shouldn't be blockers for merging this PR I wanted to highlight some common themes from my comments and areas where I think they overlapped with some of @emma-nechamkin 's questions/suggestions:

- **Separate Config Files** The design pattern that we currently follow is to define a lot of attributes about the ETL source as class globals that need to be set in each ETL instance and potentially mocked out in testing. One general question I think would be worth investigating (especially alongside the broader ETL refactor we discussed) is opportunities to standardize the attributes we set across ETL classes and define them in either a localized config/constants file or in a central config/constants file. I know there was a desire to move away from `yaml` files so I'm sure there's a good justification for avoiding this approach, but I might have missed the explanation around that design choice.
- **Canonical CBG and Census Tract Dataset** A specialized case of the approach described above is how we've hard coded a lot of attributes about the official set of CBGs and Census Tracts in the Base ETL class (e.g. number of blocks/tracts, shape of the resulting field values, etc.) This is probably fine given that these values won't change often, but I do wonder if it would be valuable to establish a canonical set of CBGs and Census Tracts (instead of hard-coding attributes about those geographic units) and use that canonical set to validate the geographic outputs of the subsequent ETL sources

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/test_etl_base.py:
```diff
@@ -0,0 +1,549 @@
+    _ETL_CLASS = ExampleETL
+
+    # In every child test class, change this to the directory of the text fixtures.
+    _DATA_DIRECTORY_FOR_TEST = (
```
Looks like we can't use `__init__` (https://stackoverflow.com/questions/62460557/cannot-collect-test-class-testmain-because-it-has-a-init-constructor-from), but I implemented a different version of this. Good idea!

lucasmbrown-usds commented on data/data-pipeline/data_pipeline/tests/sources/test_etl_base.py:
```diff
@@ -0,0 +1,549 @@
+            raise NotImplementedError("This geo level not tested yet.")
+
+    # TODO: Add a flag to make this run only when pytest is run with an argument.
+    def test_update_test_fixtures(self, mock_etl, mock_paths):
```
Hmm. I'm not sure I see what you're driving at here. The following lines: 

```
        if self._ETL_CLASS is not ExampleETL:
            raise NotImplementedError(
                "Update fixtures method not defined for this class."
            )
```

Will force all child classes to reimplement this. Does that seem sufficient to you?

emma-nechamkin commented on data/data-pipeline/data_pipeline/etl/base.py:
```diff
@@ -116,63 +124,146 @@ def transform(self) -> None:
+                "The `transform` step must set `self.output_df`."
+            )
+
+        for column_to_keep in self.COLUMNS_TO_KEEP:
```
I'm kinda with @widal001 here and kind of not -- I think canonical census tracts would be a good way to ensure we are getting a complete set of data each time, but on the other hand, data can change and so too can census tracts and having a canonical set of census tracts could be a problem?

lucasmbrown-usds replied:
Right - and to my knowledge, very few of the sources have all of the same tracts, so we'd have to have some thoughtful logic about how to decide what exactly we're comparing against the canonical tracts.

emma-nechamkin commented on data/data-pipeline/data_pipeline/etl/base.py:
```diff
@@ -1,8 +1,10 @@
+import enum
+import pathlib
```
a real nit -- since we import pathlib, can we refer to Path as pathlib.Path and then delete line 4? (I am sorry)

lucasmbrown-usds replied:
Yessss eagle eyes! I love it. 

TBH I'm always in favor of importing top-level modules and not specific items from the module. It is a little chattier, but the code is more readable because you know where that thing came from.

emma-nechamkin reviewed (commented):
i don't have much to add but this looks GREAT and I am looking forward to implementing it all!
